eth/protocols/snap/handler.go

```diff
@@ -0,0 +1,514 @@
+			if origin != (common.Hash{}) || size >= hardLimit {
+				// Request started at a non-zero hash or was capped prematurely, add
+				// the endpoint Merkle proofs
+				accTrie, err := trie.New(req.Root, backend.Chain().StateCache().TrieDB())
+				if err != nil {
+					return p2p.Send(peer.rw, storageRangesMsg, &storageRangesData{ID: req.ID})
+				}
+				var acc state.Account
+				if err := rlp.DecodeBytes(accTrie.Get(account[:]), &acc); err != nil {
+					return p2p.Send(peer.rw, storageRangesMsg, &storageRangesData{ID: req.ID})
+				}
+				stTrie, err := trie.New(acc.Root, backend.Chain().StateCache().TrieDB())
+				if err != nil {
+					return p2p.Send(peer.rw, storageRangesMsg, &storageRangesData{ID: req.ID})
+				}
+				proof := light.NewNodeSet()
+				if err := stTrie.Prove(origin[:], 0, proof); err != nil {
+					log.Warn("Failed to prove storage range", "origin", req.Origin, "err", err)
+					return p2p.Send(peer.rw, storageRangesMsg, &storageRangesData{ID: req.ID})
+				}
+				if last != (common.Hash{}) {
+					if err := stTrie.Prove(last[:], 0, proof); err != nil {
+						log.Warn("Failed to prove storage range", "last", last, "err", err)
+						return p2p.Send(peer.rw, storageRangesMsg, &storageRangesData{ID: req.ID})
+					}
+				}
+				for _, blob := range proof.NodeList() {
+					proofs = append(proofs, blob)
+				}
+				// Proof terminates the reply as proofs are only added if a node
+				// refuses to serve more data (exception when a contract fetch is
+				// finishing, but that's that).
+				break
+			}
+		}
+		// Send back anything accumulated
+		return p2p.Send(peer.rw, storageRangesMsg, &storageRangesData{
+			ID:    req.ID,
+			Slots: slots,
+			Proof: proofs,
+		})
+
+	case msg.Code == storageRangesMsg:
+		// A range of accounts arrived to one of our previous requests
+		var res storageRangesData
+		if err := msg.Decode(&res); err != nil {
+			return fmt.Errorf("%w: message %v: %v", errDecode, msg, err)
+		}
+		var (
+			keys = make([][]common.Hash, len(res.Slots))
+			vals = make([][][]byte, len(res.Slots))
+		)
+		for i, slots := range res.Slots {
+			keys[i] = make([]common.Hash, len(slots))
+			vals[i] = make([][]byte, len(slots))
+			for j, slot := range slots {
+				keys[i][j] = slot.Hash
+				vals[i][j] = slot.Body
+			}
+		}
+		return backend.OnStorage(peer, res.ID, keys, vals, res.Proof)
+
+	case msg.Code == getByteCodesMsg:
+		// Decode bytecode retrieval request
+		var req getByteCodesData
+		if err := msg.Decode(&req); err != nil {
+			return fmt.Errorf("%w: message %v: %v", errDecode, msg, err)
+		}
+		if req.Bytes > softResponseLimit {
+			req.Bytes = softResponseLimit
+		}
+		if len(req.Hashes) > maxCodeLookups {
+			req.Hashes = req.Hashes[:maxCodeLookups]
+		}
+		// Retrieve bytecodes until the packet size limit is reached
+		var (
+			codes [][]byte
+			bytes uint64
+		)
+		for _, hash := range req.Hashes {
+			if blob, err := backend.Chain().ContractCode(hash); err == nil {
```

Comment:
Following this through, down into `cachingDb`
```golang
func (db *cachingDB) ContractCode(addrHash, codeHash common.Hash) ([]byte, error) {
	if code := db.codeCache.Get(nil, codeHash.Bytes()); len(code) > 0 {
		return code, nil
	}
	code := rawdb.ReadCode(db.db.DiskDB(), codeHash)
	if len(code) > 0 {
		db.codeCache.Set(codeHash.Bytes(), code)
		db.codeSizeCache.Add(codeHash, len(code))
		return code, nil
	}
	return nil, errors.New("not found")
}
```
The underlying layers expect that the caller does _not_ request the empty-code hash, as that one hits the disk every time (avoiding the cache layers), even to the point of doing it twice: 
```
	data, _ := db.Get(hash[:])
	if len(data) != 0 {
		return data
	}
	return ReadCodeWithPrefix(db, hash)
```
So we should definitely shortcut here if the hash is `emptyCode`

 ```suggestion
			if hash == emptyCode{
				continue
			 }
			if blob, err := backend.Chain().ContractCode(hash); err == nil {
```